temporal-sdk/src/main/java/io/temporal/internal/sync/SyncWorkflowContext.java

```diff
@@ -297,45 +296,118 @@ public void invoke(Optional<Payloads> output, Failure failure) {
+      if (exception != null) {
+        runner.executeInWorkflowThread(
+            "local activity failure callback", () -> result.completeExceptionally(exception));
+      } else {
+        runner.executeInWorkflowThread(
+            "local activity completion callback", () -> result.complete(successOutput));
+      }
+    }
+  }
+
   @Override
   public <R> LocalActivityOutput<R> executeLocalActivity(LocalActivityInput<R> input) {
-    long startTime = WorkflowInternal.currentTimeMillis();
-    return new LocalActivityOutput<>(
-        WorkflowRetryerInternal.retryAsync(
-            (attempt, currentStart) -> executeLocalActivityOnce(input, attempt), 1, startTime));
+    long originalScheduledTime = System.currentTimeMillis();
+    CompletablePromise<R> result = WorkflowInternal.newCompletablePromise();
+    executeLocalActivityOverLocalRetryThreshold(input, originalScheduledTime, 1, null, result);
+    return new LocalActivityOutput<>(result);
+  }
+
+  public <R> void executeLocalActivityOverLocalRetryThreshold(
+      LocalActivityInput<R> input,
+      long originalScheduledTime,
+      int attempt,
+      @Nullable Failure previousExecutionFailure,
+      CompletablePromise<R> result) {
+    Promise<R> localExecutionResult =
+        executeLocalActivityLocally(
+            input, originalScheduledTime, attempt, previousExecutionFailure);
+
+    localExecutionResult.handle(
+        (r, e) -> {
+          if (e == null) {
+            result.complete(r);
+          } else {
+            if ((e instanceof LocalActivityCallback.LocalActivityFailedException)) {
+              LocalActivityCallback.LocalActivityFailedException laException =
+                  (LocalActivityCallback.LocalActivityFailedException) e;
+              if (laException.getBackoff() != null) {
+                WorkflowInternal.newTimer(laException.getBackoff())
+                    .thenApply(
+                        unused -> {
+                          executeLocalActivityOverLocalRetryThreshold(
+                              input,
+                              originalScheduledTime,
+                              laException.getLastAttempt() + 1,
+                              laException.getFailure(),
+                              result);
+                          return null;
+                        });
+              } else {
+                // final failure, report back
+                RuntimeException temporalFailure =
+                    FailureConverter.failureToException(
+                        laException.getFailure(), getDataConverter());
+                result.completeExceptionally(temporalFailure);
+              }
+            } else {
+              log.warn(
+                  "[BUG] executeLocalActivity() returned an exception other than LocalActivityFailedException",
+                  e);
+              result.completeExceptionally(
```

Comment:
Will this fail the workflow task?

Reply:
It will fail the workflow execution for now. I mean, this code path should never happen.